### ISSUE TITLE
Error running parallel tabu search on Windows

### ISSUE THREAD

**Issue by mcfadd** [mcfadd/Job_Shop_Schedule_Problem#4]
After building and installing [JSSP-0.1.0.win-amd64.zip](https://github.com/mcfadd/Job_Shop_Schedule_Problem/files/3413938/JSSP-0.1.0.win-amd64.zip) (Windows build), then running [TS_Example.ipynb](https://github.com/mcfadd/Job_Shop_Schedule_Problem/blob/optimization_agents/examples/TS_Example.ipynb), the following error appeared:
```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-8458fe138112> in <module>
     34                                    benchmark=benchmark,
     35                                    verbose=verbose,
---> 36                                    progress_bar=progress_bar
     37                                   )
     38 

~\Anaconda3\envs\JSSP_env\lib\site-packages\JSSP\solver.py in tabu_search_time(self, runtime, num_solutions_per_process, num_processes, tabu_list_size, neighborhood_size, neighborhood_wait, probability_change_machine, reset_threshold, initial_solutions, benchmark, verbose, progress_bar)
     70                                  probability_change_machine=probability_change_machine,
     71                                  reset_threshold=reset_threshold, initial_solutions=initial_solutions,
---> 72                                  benchmark=benchmark, verbose=verbose, progress_bar=progress_bar)
     73 
     74     def tabu_search_iter(self, iterations, num_solutions_per_process=1, num_processes=4, tabu_list_size=50,

~\Anaconda3\envs\JSSP_env\lib\site-packages\JSSP\solver.py in _tabu_search(self, stopping_condition, time_condition, num_solutions_per_process, num_processes, tabu_list_size, neighborhood_size, neighborhood_wait, probability_change_machine, reset_threshold, initial_solutions, benchmark, verbose, progress_bar)
    186         # start child processes
    187         for p in processes:
--> 188             p.start()
    189             if verbose:
    190                 print(f"child TS process started. pid = {p.pid}")

~\Anaconda3\envs\JSSP_env\lib\multiprocessing\process.py in start(self)
    103                'daemonic processes are not allowed to have children'
    104         _cleanup()
--> 105         self._popen = self._Popen(self)
    106         self._sentinel = self._popen.sentinel
    107         # Avoid a refcycle if the target function holds an indirect

~\Anaconda3\envs\JSSP_env\lib\multiprocessing\context.py in _Popen(process_obj)
    221     @staticmethod
    222     def _Popen(process_obj):
--> 223         return _default_context.get_context().Process._Popen(process_obj)
    224 
    225 class DefaultContext(BaseContext):

~\Anaconda3\envs\JSSP_env\lib\multiprocessing\context.py in _Popen(process_obj)
    320         def _Popen(process_obj):
    321             from .popen_spawn_win32 import Popen
--> 322             return Popen(process_obj)
    323 
    324     class SpawnContext(BaseContext):

~\Anaconda3\envs\JSSP_env\lib\multiprocessing\popen_spawn_win32.py in __init__(self, process_obj)
     63             try:
     64                 reduction.dump(prep_data, to_child)
---> 65                 reduction.dump(process_obj, to_child)
     66             finally:
     67                 set_spawning_popen(None)

~\Anaconda3\envs\JSSP_env\lib\multiprocessing\reduction.py in dump(obj, file, protocol)
     58 def dump(obj, file, protocol=None):
     59     '''Replacement for pickle.dump() using ForkingPickler.'''
---> 60     ForkingPickler(file, protocol).dump(obj)
     61 
     62 #

~\Anaconda3\envs\JSSP_env\lib\site-packages\JSSP\solution\makespan.cp36-win_amd64.pyd in View.MemoryView._memoryviewslice.__reduce_cython__()

TypeError: no default __reduce__ due to non-trivial __cinit__
```

The problem has to do with pickling cython memory views. When python.multiprocessing.Process starts (i.e. is spawned), the memory views in `makespan.cp36-win_amd64.pyd` cannot be pickled.

### Possible solutions

create a `__reduce__` function that converts the memory views to a numpy arrays as described in [this stackoverflow thread](https://stackoverflow.com/questions/20648167/why-cant-cython-memory-views-be-pickled)

**Comment by mcfadd** [comment]
After creating a `__reduce__` function for JSSP.solution.Solution the pickling error does not appear, however a new error appears due to the spawned child processes not being able to access the parent process's data.Data static class.

Solution's reduce function:
```python
    def __reduce__(self):
        self.machine_makespans = np.asarray(self.machine_makespans)  # need to convert memory view to np array
        return (self.__class__, (self.operation_2d_array, ))
```

New error:
```python
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\Matt\AppData\Local\Programs\Python\Python37\lib\multiprocessing\spawn.py", line 105, in spawn_main
    exitcode = _main(fd)
  File "C:\Users\Matt\AppData\Local\Programs\Python\Python37\lib\multiprocessing\spawn.py", line 115, in _main
    self = reduction.pickle.load(from_parent)
  File "C:\Users\Matt\PycharmProjects\Job_Shop_Schedule_Problem\JSSP\solution\solution.py", line 44, in __init__
    raise IncompleteSolutionException(f"Incomplete Solution of size {operation_2d_array.shape[0]}. "
JSSP.solution.solution.IncompleteSolutionException: Incomplete Solution of size 222. Should be 0
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\Matt\AppData\Local\Programs\Python\Python37\lib\multiprocessing\spawn.py", line 105, in spawn_main
    exitcode = _main(fd)
  File "C:\Users\Matt\AppData\Local\Programs\Python\Python37\lib\multiprocessing\spawn.py", line 115, in _main
    self = reduction.pickle.load(from_parent)
  File "C:\Users\Matt\PycharmProjects\Job_Shop_Schedule_Problem\JSSP\solution\solution.py", line 44, in __init__
    raise IncompleteSolutionException(f"Incomplete Solution of size {operation_2d_array.shape[0]}. "
JSSP.solution.solution.IncompleteSolutionException: Incomplete Solution of size 222. Should be 0
```

**Comment by mcfadd** [comment]
This issue was fixed in commit 5ea8f66.

The problem was solved by implementing a class that extends multiprocessing.Process and re-initializes data.Data in it's run functions before running tabu search.

I also removed Solution.__reduce__() and implemented Solution.__getstate__() and Solution.__setstate__()
